shared/shot.js

```diff
@@ -0,0 +1,726 @@
+// Note: in this library we can't use any "system" dependencies because this can be used from multiple
+// environments
+/* globals console */
+
+/** Throws an error if the condition isn't true.  Any extra arguments after the condition
+    are used as console.error() arguments. */
+function assert(condition) {
+  if (! condition) {
+    console.error.apply(console, ["Failed assertion:"].concat(Array.prototype.slice.call(arguments, 1)));
+    if (arguments.length > 1) {
+      throw new Error("Failed assertion: " + Array.prototype.slice.call(arguments, 1).join(" "));
+    } else {
+      throw new Error("Failed assertion");
+    }
+  }
+}
+
+/** True if `url` is a valid URL */
+function isUrl(url) {
+  // FIXME: this is rather naive, obviously
+  if ((/^about:.+$/i).test(url)) {
+    return true;
+  }
+  if ((/^file:\/.*$/i).test(url)) {
+    return true;
+  }
+  if ((/^data:.*$/i).test(url)) {
+    return true;
+  }
+  if ((/^chrome:.*/i).test(url)) {
+    return true;
+  }
+  if ((/^view-source:/i).test(url)) {
+    return isUrl(url.substr("view-source:".length));
+  }
+  return (/^https?:\/\/[a-z0-9\.\-]+[a-z0-9](:[0-9]+)?\/?/i).test(url);
```

Comment:
@ianb What about using https://www.npmjs.com/package/whatwg-url as a polyfill on the server? Then we could use the URL constructor, keep zero dependencies in the addon, and the server should work just fine. Maybe this could be filed as a separate issue.